Endpoint using `fileUpload` responds with a 500 status

This was discovered when testing that a 400 status is received for sending a non-multipart request body to an endpoint that is expecting multipart. See [this gist](https://gist.github.com/cruhland/7b40eef64c257e68ba6c) for a minimal test case.

The multipart parsing code in Netty throws an exception which should be converted into a 400 status by Finch, but somehow escapes. Attempting to deal with the exception using `Endpoint.handle` doesn't work, so I worked around it by defining my own version of `RequestReaders.fileUpload` which catches the exception by wrapping `req.multipart` from [this line](https://github.com/finagle/finch/blob/32f9ee812be534c1e08915998b0a95403270838c/core/src/main/scala/io/finch/RequestReaders.scala#L46) in a `scala.util.Try` instance.

I'm happy to submit a PR to fix this, but I'm not sure if my workaround is the best approach. It may be more appropriate to handle the error at a lower level so that, for example, `com.twitter.finagle.http.Request.multipart` will produce `None` instead of throwing an exception.

There are different exceptions, depending on the body contents. With `.buildPost(Buf.Utf8("&"))`, the stack trace is:

```
java.lang.IllegalArgumentException: empty name
  at org.jboss.netty.handler.codec.http.multipart.AbstractHttpData.<init>(AbstractHttpData.java:48)
  at org.jboss.netty.handler.codec.http.multipart.AbstractMemoryHttpData.<init>(AbstractMemoryHttpData.java:41)
  at org.jboss.netty.handler.codec.http.multipart.MemoryAttribute.<init>(MemoryAttribute.java:30)
  at org.jboss.netty.handler.codec.http.multipart.MixedAttribute.<init>(MixedAttribute.java:36)
  at org.jboss.netty.handler.codec.http.multipart.DefaultHttpDataFactory.createAttribute(DefaultHttpDataFactory.java:112)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.parseBodyAttributes(HttpPostStandardRequestDecoder.java:407)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.parseBody(HttpPostStandardRequestDecoder.java:225)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.<init>(HttpPostStandardRequestDecoder.java:143)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostRequestDecoder.<init>(HttpPostRequestDecoder.java:82)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostRequestDecoder.<init>(HttpPostRequestDecoder.java:56)
  at com.twitter.finagle.http.exp.Multipart$.decodeNonChunked(Multipart.scala:88)
  at com.twitter.finagle.http.Request._multipart$lzycompute(Request.scala:58)
  at com.twitter.finagle.http.Request._multipart(Request.scala:56)
  at com.twitter.finagle.http.Request.multipart(Request.scala:55)
```

With a Zip file as the body contents (my original use case), the stack trace is:

```
org.jboss.netty.handler.codec.http.multipart.HttpPostRequestDecoder$ErrorDataDecoderException: Bad string: '?% ??=�D;�����J_ ��(M�K����M-�/.�H-J�?���H���/�/�I�?�RF?�5�?? 41��� ?PK?? EH�\???u???PK????????Ln�G???????????? ???config.jsonM�;?�
�wN�hǶd�r�^�?'!�??w�"�^ R�Ų���CH��i��m�$�� '������?i�}��[��7B��k��;��?� �� �T����s�a?j�'
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.decodeAttribute(HttpPostStandardRequestDecoder.java:525)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.setFinalBuffer(HttpPostStandardRequestDecoder.java:503)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.parseBodyAttributes(HttpPostStandardRequestDecoder.java:419)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.parseBody(HttpPostStandardRequestDecoder.java:225)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.<init>(HttpPostStandardRequestDecoder.java:143)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostRequestDecoder.<init>(HttpPostRequestDecoder.java:82)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostRequestDecoder.<init>(HttpPostRequestDecoder.java:56)
  at com.twitter.finagle.http.exp.Multipart$.decodeNonChunked(Multipart.scala:88)
  at com.twitter.finagle.http.Request._multipart$lzycompute(Request.scala:58)
  at com.twitter.finagle.http.Request._multipart(Request.scala:56)
  at com.twitter.finagle.http.Request.multipart(Request.scala:55)
  ... 43 elided
Caused by: java.lang.IllegalArgumentException: URLDecoder: Illegal hex characters in escape (%) pattern - For input string: " ?"
  at java.net.URLDecoder.decode(URLDecoder.java:194)
  at org.jboss.netty.handler.codec.http.multipart.HttpPostStandardRequestDecoder.decodeAttribute(HttpPostStandardRequestDecoder.java:521)
  ... 53 more
```

The general problem seems to be that Netty's `HttpPostRequestDecoder` can throw exceptions on arbitrary (hence non-multipart) input.

--- a/core/src/main/scala/io/finch/RequestReaders.scala
+++ b/core/src/main/scala/io/finch/RequestReaders.scala
@@ -4,7 +4,7 @@ import com.twitter.finagle.http.{Cookie, Request}
 import com.twitter.finagle.http.exp.Multipart.FileUpload
 import com.twitter.finagle.netty3.ChannelBufferBuf
 import com.twitter.io.{Buf, Charsets}
-import com.twitter.util.Future
+import com.twitter.util.{Future, Try}
 
 trait RequestReaders {
 
@@ -43,7 +43,7 @@ trait RequestReaders {
     req.cookies.get(cookie)
 
   private[finch] def requestUpload(upload: String)(req: Request): Option[FileUpload] =
-    req.multipart.flatMap(m => m.files.get(upload).flatMap(fs => fs.headOption))
+    Try(req.multipart).getOrElse(None).flatMap(m => m.files.get(upload).flatMap(fs => fs.headOption))
 
   // A convenient method for internal needs.
   private[finch] def rr[A](i: RequestItem)(f: Request => A): RequestReader[A] =

--- a/core/src/test/scala/io/finch/MultipartParamSpec.scala
+++ b/core/src/test/scala/io/finch/MultipartParamSpec.scala
@@ -44,6 +44,12 @@ class MultipartParamSpec extends FlatSpec with Matchers {
     buf.length should be > 0
   }
 
+  it should "produce an error if the request body is not parsable as multipart" in {
+    val request = RequestBuilder().url("http://example.com").buildPost(Buf.Utf8("&"))
+    val futureResult = fileUpload("notmultipart")(request)
+    an [Error.NotPresent] shouldBe thrownBy(Await.result(futureResult))
+  }
+
   "An OptionalMultipartFile" should "have a filename if it exists" in {
     val request = requestFromBinaryFile("/upload.bytes")
     val futureResult: Future[Option[Multipart.FileUpload]] = fileUploadOption("groups")(request)
@@ -53,6 +59,12 @@ class MultipartParamSpec extends FlatSpec with Matchers {
   it should "be empty when the upload name exists but is not an upload" in {
     val request = RequestBuilder().url("http://localhost/").addFormElement("groups" -> "foo").buildFormPost()
     val futureResult: Future[Option[Multipart.FileUpload]] = fileUploadOption("groups")(request)
+    Await.result(futureResult) shouldBe None
+  }
+
+  it should "produce an error if the request body is not parsable as multipart" in {
+    val request = RequestBuilder().url("http://example.com").buildPost(Buf.Utf8("&"))
+    val futureResult = fileUploadOption("notmultipart")(request)
     Await.result(futureResult) shouldBe None
   }
 

